Проект. Покрытие приложения тестами

EldenLordCommander opened the pull request:
Здравствуйте! Выкладывают текущий вариант работ по спринту 6. В рамках этого спринта были проделаны следующие работы, основываясь на теории и примерах из практикума: 
- покрыл тестами все редьюсеры - в папке reducers рядом с каждым редьюсером есть файл теста
- в каждом файле теста редьюсера проверяется начальное состояние и все обработчики текущего редьюсера
- установил все требуемые для тестирования библиотеки в проект
- в cypress описал тесты для проверки открытия и закрытия модального окна ингредиента из конструктора, переноса ингридиентов в конструктор, логина, создания заказа, открытия и закрытия окна с номером заказа
- опубликовал проект на gh-pages в соответствии с руководством в Практикуме, ссылку добавил в начало readme-файла

Буду рад, если проверите и подскажите, что требует доработки и исправления для принятия в рамках данного спринта.
Заранее спасибо!

EldenLordCommander committed:
Проект. Покрытие приложения тестами

- покрыл тестами все редьюсеры
- в каждый тесте проверяется начальное состояние и все обработчики
- установил все требуемые для тестирования библиотека
- в cypress описал тесты для проверки открытия и закрытия модального окна ингредиента из конструктора, переноса ингридиентов в конструктор, логина, создания заказа, открытия и закрытия окна с номером заказа
- опубликовал проект на gh-pages

gennady-bars commented on src/utils/burger-api.ts:
```diff
@@ -156,7 +156,7 @@ export const login = (form : TMainForm) => {
     })
         .then(checkResponse)
```
**Можно лучше**

Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request`. Все остальное будет без изменений. Код станет чище

gennady-bars commented on src/utils/burger-api.ts:
```diff
@@ -215,6 +215,7 @@ export const refreshTokenRequest = () => {
 
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/services/reducers/wsReducer.test.js:
```diff
@@ -0,0 +1,81 @@
+            }
+        )
+    });
+}) 
```
Если есть **инпуты**, то они должны быть внутри тега `form`, и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.

gennady-bars replied:
`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

gennady-bars commented on src/services/reducers/wsReducer.test.js:
```diff
@@ -0,0 +1,81 @@
+            }
+        )
+    });
+
```
` onClick={(e)=>{login(e)}}` в логине и других страницах

это нужно удалить с кнопки сабмита


**Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**

gennady-bars commented on src/services/reducers/wsReducer.ts:
```diff
@@ -13,7 +13,7 @@ export type TWSReducerState = {
     orders: TOrderRequest;
 }
 
```
**Можно лучше**

Можно сделать `ProtectedRoute` примерно так:

```jsx
const ProtectedRoute = ({ onlyForAuth, children, ...rest }) => {
  const isAuthorized = getCookie("accessToken");
  const location = useLocation();

  if (!onlyForAuth && isAuthorized) {
    const { from } = location.state || { from: { pathname: "/" } };
    return (
      <Route {...rest}>
        <Redirect to={from} />
      </Route>
    );
  }

  if (onlyForAuth && !isAuthorized) {
    return (
      <Route {...rest}>
        <Redirect to={{ pathname: "/login", state: { from: location } }} />
      </Route>
    );
  }

  return <Route {...rest}>{children}</Route>;
};

```

Этот рут может и переадресовывать пользователя туда, откуда он пришел, если он уже авторизован, тогда  не нужно будет дублировать переадресацию на страницах Логина, Регистрации, Восстановления пароля и тд

gennady-bars commented on src/utils/burger-api.ts:
```diff
@@ -223,6 +224,7 @@ export const fetchWithRefresh = async () => {
             }
```
Вам нужно добавить скрипт запуска `cypress` тестов ` "cypress": "cypress open cache clear",` в `package.json`

gennady-bars commented on cypress/e2e/burger-constructor.cy.ts:
```diff
@@ -0,0 +1,51 @@
+
+describe('Проект работает', function () {
+  before(function () {
+    cy.visit('http://localhost:3000');
```
`http://localhost:3000` повторяется несколько раз в коде тестов. Нужно вынести в константу `testUrl`, чтобы можно было одним движением изменить урл для тестов
А можно в документации посмотреть, как настроить быстро `baseUrl` для cypress https://docs.cypress.io/api/commands/visit#Visit-is-automatically-prefixed-with-baseUrl

gennady-bars commented on cypress/e2e/burger-constructor.cy.ts:
```diff
@@ -0,0 +1,51 @@
+    cy.visit('http://localhost:3000');
+
+    cy.get('#60d3b41abdacab0026a733c7').click();
+    cy.get('#modal_form').contains('Детали ингридиента');
```
`'#modal_form'` повторяется 15 раз в тестах

gennady-bars replied:
Если **строка-селектор** повторяется более 2х раз в коде, значит, этот селектор нужно выносить в **константу**, чтобы можно было одним движением изменить его и не искать по всему коду дублирования этого селектора.

gennady-bars replied:
либо нужно пользоваться переменными в `cypress`. Ищется 1 раз, а дальше в цепочке используется переменная через `@` (алиасы)


Подробнее https://www.software-testing.ru/library/testing/testing-tools/3758-cypress-basics-variables

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная

* Приложение задеплоено
* Все юнит-тесты (51 шт) проходят успешно
* Все cypress-тесты проходят успешно
* Код тестов аккуратный и понятный

но есть некоторые недочеты: 

* Скрин https://disk.yandex.ru/i/szKBuZCjwj6i9A трубно читается тест темно-синим на черном фоне. Лучше сделать по макету в фигме (белым)
* Если есть **инпуты**, то они должны быть внутри тега `form`, и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.
* **Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**
* `Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.
* `data.js` нужно тоже в `ts` переделать
* в **истории** ничего не отображается https://disk.yandex.ru/i/FPZVHbtvaKADFA 
* нет автологина
* Если пользователь **не авторизован** и пытается попасть на защищённый маршрут — его **переадресовывает** на маршрут `/login`. После авторизации пользователь автоматически переадресовывается на тот маршрут, к которому у него **не получилось** получить доступ ранее. Это делается с помощью ` return <Redirect to={location?.state?.from || '/'} />;` (эти данные записываются в `ProtectedRoute`, а используются в `Login` для `Redirect`)
* Вам нужно добавить скрипт запуска `cypress` тестов ` "cypress": "cypress open cache clear",` в `package.json`
* `http://localhost:3000` повторяется несколько раз в коде тестов. Нужно вынести в константу `testUrl`, чтобы можно было одним движением изменить урл для тестов
А можно в документации посмотреть, как настроить быстро `baseUrl` для cypress https://docs.cypress.io/api/commands/visit#Visit-is-automatically-prefixed-with-baseUrl 
* `'#modal_form'` повторяется 15 раз в тестах
* Если **строка-селектор** повторяется более 2х раз в коде, значит, этот селектор нужно выносить в **константу**, чтобы можно было одним движением изменить его и не искать по всему коду дублирования этого селектора.
* Вам нужно удалить `DEMO`-тесты cypress. Скрин https://disk.yandex.ru/i/oBF1Sjf702P7cw . Только свой нужно оставить по бургеру





**Можно лучше**

*  при клике на **логотип** обычно идет переход на главную страницу
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа.
*  желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

EldenLordCommander committed:
fix недочетов

- удалил ненужные тесты и файлы
- на страницах с формами перевел все на form с submit
- из модалки убрал вызов экшенов, немного поменял механизм работы модалки
- поправил автологин и редирект после авторизации
- добавил скрипт запуска тестов
- вынес адрес из теста в конфиг
- в тесте вынес повторяющийся селектор в переменную

EldenLordCommander commented:
Здравствуйте!
Внес изменения в проект, исправив указанные недочеты:
-  поменял шрифт у элементов, чтобы они лучше читались
- на pages с инпутами вынес все в формы с сабмитом
- поправил модал: убрал оттуда экшены, немного переделал механизм работы открытия и закрытия и очистки хранилища
-  удалил data.js - файл устарел и не используется
- не смог воспроизвести ошибку, видимо в процессе фикса остальных проблем решилась и эта, у меня история открывается и грузится корректно постоянно
- поправил автологин
- исправил работу редиректа со страницы логина при попытке попасть на защищенный маршрут. После логина возвращает на страницу, на которой юзер был
- добавил скрипт в конфиг cypress
- вынес адрес сайта в конфиг
- вынес '#modal_form' в переменную
-  удалил тесты созданные по умолчанию

Буду рад, если проверите текущий билд, и укажете, все ли корректно исправлено. И, если можно, подробнее указать, об моментах, которые не работают, как с историей, если повторится.
Заранее спасибо

gennady-bars commented:
> моментах, которые не работают, как с историей, если повторится.

теперь все хорошо в истории

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.



* лучше теперь удалить все неиспользуемые импорты и переменные. Скрин https://disk.yandex.ru/i/oKcEMnyeSfA9DA . Их много в каждом файле


Удачи в Вашей новой профессии.

**Несколько советов для дальнейшего развития:**

* сверстайте несколько небольших проектов для себя, для закрепления знаний, и чтобы можно добавить их в портфолио
* ходите на собеседования: опыт прохождения собеседований не менее важен знаний теории, и Вы поймете, какие знания требуются на рынке, и где Вы проседаете
* читайте [habr](https://habr.com/ru/) и слушаете подкаксты: они очень хорошо расширяют кругозор и позволяют узнать, что происходит в отрасли:

https://soundcloud.com/everyonecan

https://soundcloud.com/podlodka

https://soundcloud.com/devschacht

https://soundcloud.com/begebot

https://soundcloud.com/web-standards

https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw
